include/openssl/ssl.h

```diff
@@ -1395,6 +1395,31 @@ __owur int SSL_CIPHER_get_bits(const SSL_CIPHER *c, int *alg_bits);
 __owur char *SSL_CIPHER_get_version(const SSL_CIPHER *c);
 __owur const char *SSL_CIPHER_get_name(const SSL_CIPHER *c);
 __owur uint32_t SSL_CIPHER_get_id(const SSL_CIPHER *c);
+typedef enum {
+    SSL_CIPHER_KX_RSA = 1,
+    SSL_CIPHER_KX_ECDHE,
+    SSL_CIPHER_KX_DHE,
+    SSL_CIPHER_KX_ECDHE_PSK,
+    SSL_CIPHER_KX_DHE_PSK,
+    SSL_CIPHER_KX_RSA_PSK,
+    SSL_CIPHER_KX_PSK,
+    SSL_CIPHER_KX_SRP,
+    SSL_CIPHER_KX_GOST
```

Comment:
Ok, let's try the NID route. This is consistent with what we already have for cipher/digest, and has the added bonus that you get human-readable representation for free.

I'd suggest (but my ideas aren't very mature at this point) adding new NID_tls_kx_foo and NID_tls_auth_foo IDs for consistency. I can't find many that we can reuse without creating confusion. The RSA ones, perhaps, could be aliased to existing ones, e.g. NID_tls_*_rsa could be aliased to NID_rsa. And in some cases you can alias them to each other, e.g. NID_tls_kx_psk -> NID_tls_auth_psk perhaps. These don't correspond to object identifiers anyway.

Apparently the way to add them is to add them to crypto/objects/objects.txt. Give it a try.

(Man, the TLS way of specifying ciphers is a mess.)